### ISSUE TITLE
Mediator pipeline behavior for specific commands - the comeback

### ISSUE THREAD

**Issue by Lobosque** [jbogard/MediatR#562]
After the epic journey for getting https://github.com/dotnet/runtime/pull/39540 merged, is it possible to solve these issues in the current .net 5 RC?

https://github.com/jbogard/MediatR/issues/542
https://github.com/jbogard/MediatR/issues/284

**Comment by lilasquared** [comment]
Yep, you just register the behaviors as you normally would.

```csharp
// Startup.cs
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(PingBehavior<,>));
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(PongBehavior<,>));

// PingBehavior.cs
public class PingBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : Ping
{
    public Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
    {
        Debug.WriteLine("PingBehavior");
        return next();
    }
}

// PongBehavior.cs
public class PongBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : Pong
{
    public Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
    {
        Debug.WriteLine("PongBehavior");
        return next();
    }
}
```

**Comment by rohansp** [comment]
@Lobosque @lilasquared Is anyone working on this or can I take this?

**Comment by Lobosque** [comment]
@rohansp please go ahead, I am not doing any work related to this issue.

**Comment by lilasquared** [comment]
what is there to work on?

**Comment by turgay-ozgur** [comment]
I am using .NET 5.0 RC2. 

There is no exception with that but still it is not working. The `Handle` method of that pipeline behavior not being called if I wrap the `TResponse` like `Result<TResponse>`. Am I missing something?

```csharp
public class ValidationBehaviour<TRequest, TResponse> : IPipelineBehavior<TRequest, Result<TResponse>>
        where TRequest : IRequest<Result<TResponse>>
{
    public async Task<Result<TResponse>> Handle(TRequest request, CancellationToken cancellationToken,
            RequestHandlerDelegate<Result<TResponse>> next)
    {
        return new Result<TResponse>();
    }
}
```

**Comment by lilasquared** [comment]
I don't know what you  mean by "wrap TResponse with Result<TResponse>" can you give the example request / response objects you're working with?  Also can you share your registration code?

**Comment by turgay-ozgur** [comment]
Sure @lilasquared, thank you for your help.

```csharp
public class FooRequest : IRequest<Result<FooDto>>
{
}

public class FooRequestHandler : IRequestHandler<FooRequest, Result<FooDto>>
{
}
```

**Comment by lilasquared** [comment]
Okay I thought that is what you meant just wanted to be sure.  I think the problem is the use of `Result<TResponse>` in the definition of the class interface, though I am not in a place to test that theory.  We configured ours with the result envelopes implementing an interface and then using that interface in the constraint on the result rather than the request.  You do lose access to the type of the response object that way, but we build results using Activator and reflection for those cases. 

```csharp
    public class ExceptionBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
        where TResponse : class, IResult
    {
        public async Task<TResponse> Handle(TRequest request,
                                            CancellationToken cancellationToken,
                                            RequestHandlerDelegate<TResponse> next)
        {
            //...
        }
    }
```

**Comment by turgay-ozgur** [comment]
Activator and reflection may help but those come with performance problems. Compiled lambda expressions more performant to create new instances but adding extra complexity. I am trying to handle validations without exception. But seems like I have to throw a validation exception anyway.

**Comment by lilasquared** [comment]
We handle validation without explicitly throwing any exception and just returning a failure result

**Comment by turgay-ozgur** [comment]
Yes, it seems possible when using activator or using `IResult` instead of `IResult<>` as "TResponse".

**Comment by turgay-ozgur** [comment]
Here is your answer for the same purpose. [#466](https://github.com/jbogard/MediatR/issues/466#issuecomment-561874242).

This was not working because the open generics not being handled on dotnet's IoC implementation before. So, why it is still not working even the open generics handled on .NET 5.0 RC?

**Comment by lilasquared** [comment]
That example uses a different DI, and is not using a constraint on the generic logging behavior. If you keep reading down I was mistaken and it was the constrained generic that was not working previously in dotnet core

**Comment by turgay-ozgur** [comment]
Just for anyone who need to create instance for that kind of requirements the below code may help to performant instance creation. Only supports parameterless constructors but can be extended to cover ctor parameters.

```csharp
public static class New<T>
{
    public static readonly Func<T> Instance = Expression.Lambda<Func<T>>(Expression.New(typeof(T))).Compile();
}
```

**Comment by visar-elmazi-allocate** [comment]
Following up on @lilasquared  
You can also use a Flag interface i.e. IPersistent

`public interface IPersistent { }`

to flag all your requests for which you want to apply the behavior:

```
public class PersistentBehavior<TRequest, TResponse>
    : IPipelineBehavior<TRequest, TResponse> where TRequest: **IPersistent**  {
    ..
}
```

If you don't mind using the flagging, this should give you somewhat more freedom to chose what behavior applies where.

**Comment by lonix1** [comment]
### TL;DR for multiple pipelines

Register pipelines:
````cs
// pipeline 1's behaviors
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(FooBehavior<,>));
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(BarBehavior<,>));
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(BazBehavior<,>));

// pipeline 2's behaviors
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(CatBehavior<,>));
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(DogBehavior<,>));
services.AddScoped(typeof(IPipelineBehavior<,>), typeof(PigBehavior<,>));
````

Define pipelines:
````cs
// pipeline 1
public interface IPipeline1 { }
public class FooBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IPipeline1 { }
public class BarBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IPipeline1 { }
public class BazBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IPipeline1 { }

// pipeline 2
public interface IPipeline2 { }
public class CatBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IPipeline2 { }
public class DogBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IPipeline2 { }
public class PigBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IPipeline2 { }
````

Define requests:
````cs
// requests to be processed by pipeline 1
public class ARequest : IRequest, IPipeline1 { }
public class BRequest : IRequest, IPipeline1 { }
public class CRequest : IRequest, IPipeline1 { }

// requests to be processed by pipeline 2
public class XRequest : IRequest, IPipeline2 { }
public class YRequest : IRequest, IPipeline2 { }
public class ZRequest : IRequest, IPipeline2 { }
````

**Comment by jbogard** [comment]
^^ This is the way.

**Comment by JFercan** [comment]
@lonix1 Does you solution work with IRequest<T>?  I tried and kept getting " violates the constraint of type TRequest"  exception.  I had something like:

```
public class ARequest : IRequest<Result<int>>, IPipeline1 { }

public class ARequest : IRequest<Result<int>>, IPipeline2 { }
```

And tried both of these for the specific behaviours:

```
public class FooBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IPipeline1 { }

public class FooBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> where TRequest : IRequest<TResponse>, IPipeline1 { }
```

Thank you
